doc/design/secret-provisioning.md

```diff
@@ -0,0 +1,128 @@
+client library while it generates attestation evidences, and by the server library
+when it verifies such evidences. Higher number indicates more abstraction in the
+attestation model.
+
+* **model 1**: The client generates a key pair, and an SGX report with hash of
+the public key being the report data. The public key (but not the private key)
+and the SGX report are sent to the server. The server will verify the report's
+validity, verify the public key hash, and consider the enclave identity while
+deciding whether to release the secret or not.
+
+* **model 2**: The client generates a self-signed x509 certificate with
+embedded SGX report. The hash of the public key in the certificate is the
+report data. The certificate is sent to the server. The server will verify the
+report's validity, verify the public key hash, and consider the enclave
+identity while deciding whether to release the secret or not.
+
+* **model 3**: The client generates attestation evidences necessary to obtain
+a token from an attestation service, such as MAA. The token is sent to the
+server. The server will verify the validity of the token, and verify various
+claims within the token, including but not limited to public key and TEE
+identity, while deciding whether to release the secret or not.
+
+## The client library
+
+Mystikos will provide a simple reference implementation of the client library
+that exposes the following APIs to be called by Mystikos runtime while
+performing secret provisioning:
+
+* `generate_attestation_evidence_with_keys` (for attestation model 1)
+* `generate_x509_cert_with_attestation_extension` (for attestation model 2)
+* `generate_attestation_service_token` (for attestation model 3)
+* `unwrap_secret` (for all attestation models)
+
+A client library could choose to implement only 1 or 2 `generate_xxx` functions
+for the specific attestation models supported on the server side. For example,
+the client library we provide for Azure M-HSM only implements
+`generate_attestation_service_token` (attestation model 3)
+because that's the only model M-HSM service supports.
+
+The simple reference implementation only supports the minimum requirement for
+a secret provisioning process. Any additional security enhancements, such as
+using nonce or signing and verifying the response from the server could be
+implemented in custom client/server libraries as long as the enhancement
+is agreed on by both the client and the server, and the implementation
+fulfills the same set of APIs.
+
+Users can plug in their own client library by including the library in rootfs
+and providing the corresponding `ClientLibPath` in config.json.
+
+## The server library
+
+Mystikos will provide a simple reference implementation of the server library
+that exposes the following APIs to be called by the service that releases the
+secrets:
+
+* `verify_attestation_evidence_with_key` (for attestation model 1)
```

Comment:
Will the implementation of the APIs for model 1 and 2 directly base on OE's verify APIs? If so, this implies that the server takes hard dependency on the DCAP library.  Or will the implementation supports a model where the server talks remote verifier (e.g., MAA) to validate the SGX report first?

Reply:
I clarified in the update. There are options that should be weighed by the users.